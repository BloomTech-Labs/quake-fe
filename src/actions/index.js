import axios from "axios";

// quakeReducer Actions
export const QUAKE_FETCH = "QUAKE_FETCH";
export const QUAKE_FETCH_ERROR = "QUAKE_FETCH_ERROR";
export const DISPLAY_QUAKES = "DISPLAY_QUAKES ";
export const SORT_QUAKES = "SORT_QUAKES";
export const UPDATE_VIEWPORT = "UPDATE_VIEWPORT";

// searchReducer Actions
export const UPDATE_SEARCH_PARAMS = "UPDATE_SEARCH_PARAMS";

export const quakeFetch = (theQuery) => (dispatch) => {
  dispatch({ type: QUAKE_FETCH });
  console.log("Query to USGS: ", theQuery);
  setTimeout(() => {
    axios
      .get(
        `https://earthquake.usgs.gov/fdsnws/event/1/query?format=geojson&limit=100${theQuery}`
      )
      .then((res) => {
        console.log(res.data.features);
        dispatch({ type: DISPLAY_QUAKES, quakeData: res.data.features });
      })
      .catch((err) => {
        console.log("Error Fetching Earthquake Data From USGS: ", err);
        dispatch({ type: QUAKE_FETCH_ERROR });
      });
  }, 500);
};

export const updateSearchParams = (theParams) => (dispatch) => {
  dispatch({ type: UPDATE_SEARCH_PARAMS, params: theParams });
};

export const quakeSort = (sortBy, quakes) => (dispatch) => {
  if (sortBy === "ascending magnitude") {
    const sortedQuakes = quakes.sort(
      (a, b) => a.properties.mag - b.properties.mag
    );

    console.log("SORTED QUAKES: ", sortedQuakes);
    dispatch({ type: SORT_QUAKES, quakeData: [] });
    dispatch({ type: SORT_QUAKES, quakeData: sortedQuakes });
  } else if (sortBy === "descending magnitude") {
    const sortedQuakes = quakes.sort(
      (a, b) => b.properties.mag - a.properties.mag
    );

    console.log("SORTED QUAKES: ", sortedQuakes);
    dispatch({ type: SORT_QUAKES, quakeData: [] });
    dispatch({ type: SORT_QUAKES, quakeData: sortedQuakes });
  } else if (sortBy === "newest") {
    const sortedQuakes = quakes.sort(
      (a, b) => b.properties.time - a.properties.time
    );

    console.log("SORTED QUAKES: ", sortedQuakes);
    dispatch({ type: SORT_QUAKES, quakeData: [] });
    dispatch({ type: SORT_QUAKES, quakeData: sortedQuakes });
  } else if (sortBy === "oldest") {
    const sortedQuakes = quakes.sort(
      (a, b) => a.properties.time - b.properties.time
    );

    console.log("SORTED QUAKES: ", sortedQuakes);
    dispatch({ type: SORT_QUAKES, quakeData: [] });
    dispatch({ type: SORT_QUAKES, quakeData: sortedQuakes });
  }
};

export const setViewport = (viewport) => (dispatch) => {
  console.log('UPDATING VIEWPORT TO: ', viewport);
    let newVP = {
      ...viewport,
      width: "100%",
      height: "100%",
    };
    dispatch({ type: UPDATE_VIEWPORT, viewportInfo: newVP });
  
  
}