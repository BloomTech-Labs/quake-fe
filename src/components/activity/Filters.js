import React, { useEffect, useRef } from "react";
import { connect } from "react-redux";
import { quakeFetch, updateSearchParams } from "../../actions";
// import SearchBar from "./search/SearchBar";
import useDarkMode from "../../customHooks/useDarkMode";

function Filters({
  quakeFetch,
  updateSearchParams,
  maxradiuskm,
  starttime,
  endtime,
  minmagnitude,
  maxmagnitude,
  latitude,
  longitude,
}) {
  // Toggles Dark Mode CSS theme. See src/customHooks/useDarkMode.js
  const [darkMode, setDarkMode] = useDarkMode("dark-mode");

  const toggleDarkMode = (e) => {
    e.preventDefault();
    setDarkMode(!darkMode);
  };

  // Set Ref for off click to close search.
  const searchRef = useRef(null);

  function onClickRef(e) {
    const isOutside = searchRef.current.contains(e.target);

    if (isOutside === false) {
      toggleSearch();
    }
  }

  // The query parameters to be sent to USGS. Updates with state changes
  const USGSQuery = `&starttime=${starttime}&endtime=${endtime}&minmagnitude=${minmagnitude}&maxmagnitude=${maxmagnitude}&maxradiuskm=${maxradiuskm}&latitude=${latitude}&longitude=${longitude}`;

  // Initial Quake Search, runs on first load
  useEffect(() => {
    quakeFetch(USGSQuery);
  }, []);

  // dispatches quakeFetch actions with the query for USGS upon form submit
  const formSubmitCallback = (e) => {
    e.preventDefault();
    toggleSearch();
    quakeFetch(USGSQuery);
  };

  // Updates the search state dynamically for each input due to shared key names
  const handleChanges = (e) => {
    e.preventDefault();

    updateSearchParams({
      name: e.target.name,
      value: e.target.value,
    });
  };

  const toggleSearch = () => {
    // Toggles visibility of search menu
    const searchMenu = document.getElementById("search-menu");
    searchMenu.style.display === "block"
      ? (searchMenu.style.display = "none")
      : (searchMenu.style.display = "block");
  };

  return (
    <dialog id="search-menu" onClick={onClickRef} className="search-menu">
      <form ref={searchRef} onSubmit={formSubmitCallback} name="filters-form">
        <fieldset id="advanced-filters" className="advanced-filters">
          <legend>Advanced Filters</legend>

          <fieldset className="radius-field">
            <legend>Search Radius:</legend>

            <label htmlFor="kilometers">
              Kilometers
              <input
                id="kilometers"
                type="number"
                name="maxradiuskm"
                onChange={handleChanges}
                value={maxradiuskm}
                min="1"
                max="20001.6"
                step="0.1"
              />
            </label>
          </fieldset>

          <fieldset className="date-field">
            <legend>Date Range:</legend>

            <label htmlFor="start-date">
              Start
              <input
                id="start-date"
                type="date"
                name="starttime"
                onChange={handleChanges}
                value={starttime}
              />
            </label>

            <label htmlFor="end-date">
              End
              <input
                id="end-date"
                type="date"
                name="endtime"
                onChange={handleChanges}
                value={endtime}
              />
            </label>
          </fieldset>

          <fieldset className="mag-field">
            <legend>Magnitude Range:</legend>

            <label htmlFor="start-magnitude">
              Start
              <input
                id="start-magnitude"
                type="number"
                name="minmagnitude"
                min="0"
                max="10"
                step="0.1"
                onChange={handleChanges}
                value={minmagnitude}
              />
            </label>

            <label htmlFor="end-magnitude">
              End
              <input
                id="end-magnitude"
                type="number"
                name="maxmagnitude"
                min="0"
                max="10"
                step="0.1"
                onChange={handleChanges}
                value={maxmagnitude}
              />
            </label>
          </fieldset>
        </fieldset>

        <button type="submit" className="search-submit-button">
          Search For Activity
        </button>

        <p
          type="button"
          role="button"
          className="close-sort"
          onClick={toggleSearch}
        >
          Close
        </p>

        <aside className="dark-mode-toggle">
          <div
            aria-label="dark mode toggle"
            type="button"
            role="switch"
            aria-checked={false}
            onClick={toggleDarkMode}
            className={darkMode ? "toggle-switch toggled" : "toggle-switch"}
          />
        </aside>
        <p className="dark-mode-text">Dark mode</p>
      </form>
    </dialog>
  );
}

const mapStateToProps = (state) => {
  return {
    starttime: state.searchReducer.starttime,
    endtime: state.searchReducer.endtime,
    minmagnitude: state.searchReducer.minmagnitude,
    maxmagnitude: state.searchReducer.maxmagnitude,
    maxradiuskm: state.searchReducer.maxradiuskm,
    latitude: state.searchReducer.latitude,
    longitude: state.searchReducer.longitude,
  };
};

export default connect(mapStateToProps, {
  quakeFetch,
  updateSearchParams,
})(Filters);
