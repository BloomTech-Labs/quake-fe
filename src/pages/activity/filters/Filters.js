import React, { useEffect, useRef } from "react";
import { connect } from "react-redux";
import { firstLoad, quakeFetch, updateSearchParams } from "../../../redux/actions";
import QuickFilters from './quickFilters'
import RadiusFilter from './radiusFilter'
import DateFilter from "./dateFilter";
import MagFilter from "./magFilter";
 
function Filters({
 quakeFetch,
 firstLoad,
 updateSearchParams,
 maxradiuskm,
 starttime,
 endtime,
 minmagnitude,
 maxmagnitude,
 latitude,
 longitude,
}) {
 // Set Ref for off click to close search.
 const searchRef = useRef(null);
 
 function onClickRef(e) {
   const isOutside = searchRef.current.contains(e.target);
 
   if (isOutside === false) {
     toggleSearch();
   }
 }
 
 // The query parameters to be sent to USGS. Updates with state changes
 const USGSQuery = `&starttime=${starttime}&endtime=${endtime}&minmagnitude=${minmagnitude}&maxmagnitude=${maxmagnitude}&maxradiuskm=${maxradiuskm}&latitude=${latitude}&longitude=${longitude}`;
 const firstLoadQuery = "https://quakelabs-be-production.herokuapp.com/api/activity/first-load";
 
 // Initial Quake Search, runs on first load
 useEffect(() => {
   firstLoad(firstLoadQuery, USGSQuery, quakeFetch);
 }, []);


  // dispatches quakeFetch actions with the query for USGS upon form submit
  const formSubmitCallback = (e) => {
    e.preventDefault();
    quakeFetch(USGSQuery);
  };

  // quick filters
  const quickFilters = (e) => {
    e.preventDefault();
    console.log('request coming?', e.target.name)
    firstLoad(e.target.name, USGSQuery, quakeFetch);
    toggleSearch();
  }

  // Updates the search state dynamically for each input due to shared key names
  const handleChanges = (e) => {
    e.preventDefault();

    updateSearchParams({
      name: e.target.name,
      value: e.target.value,
    });
  };

  const toggleSearch = () => {
    // Toggles visibility of search menu
    const searchMenu = document.getElementById("search-menu");
    searchMenu.style.display === "flex"
      ? (searchMenu.style.display = "none")
      : (searchMenu.style.display = "flex");
  };

  return (
    <dialog id="search-menu" onClick={onClickRef} className="search-menu">
      <form ref={searchRef} onSubmit={formSubmitCallback} name="filters-form">
        <QuickFilters quickFilters={quickFilters}/>
        <fieldset id="advanced-filters" className="advanced-filters">
          <legend id='advanced-filters-label'>Advanced Filters</legend>
          <RadiusFilter handleChanges={handleChanges} maxradiuskm={maxradiuskm}/>
          <DateFilter handleChanges={handleChanges} starttime={starttime} endtime={endtime}/>
          <MagFilter handleChanges={handleChanges} minmagnitude={minmagnitude} maxmagnitude={maxmagnitude}/>
        </fieldset>

        <button type="submit" onClick={toggleSearch} className="search-submit-button">
          Search For Activity
        </button>

        <p type="button" role="button" className="close-sort" onClick={toggleSearch}>
          Close
        </p>
        
      </form>
    </dialog>
  );
}

const mapStateToProps = (state) => {
  return {
    starttime: state.searchReducer.starttime,
    endtime: state.searchReducer.endtime,
    minmagnitude: state.searchReducer.minmagnitude,
    maxmagnitude: state.searchReducer.maxmagnitude,
    maxradiuskm: state.searchReducer.maxradiuskm,
    latitude: state.searchReducer.latitude,
    longitude: state.searchReducer.longitude,
  };
};

export default connect(mapStateToProps, {
  firstLoad,
  quakeFetch,
  updateSearchParams,
})(Filters);
