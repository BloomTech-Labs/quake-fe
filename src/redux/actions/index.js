/* eslint-disable no-eval */
import axios from "axios";
import * as turf from "@turf/turf";
import { initialSearchState } from "../reducers/searchReducer";

// quakeReducer Actions
export const QUAKE_FETCH = "QUAKE_FETCH";
export const QUAKE_FETCH_ERROR = "QUAKE_FETCH_ERROR";
export const DISPLAY_QUAKES = "DISPLAY_QUAKES ";
export const SORT_QUAKES = "SORT_QUAKES";

// searchReducer Actions
export const UPDATE_SEARCH_PARAMS = "UPDATE_SEARCH_PARAMS";

// mapReducer Actions
export const UPDATE_VIEWPORT = "UPDATE_VIEWPORT";
export const JUMP_VIEWPORT = "JUMP_VIEWPORT";

// newsReducer Actions
export const NEWS_FETCH = "NEWS_FETCH";
export const NEWS_FETCH_ERROR = "NEWS_FETCH_ERROR";
export const DISPLAY_NEWS = "DISPLAY_NEWS";

let sortedQuakes,
  sortInfo = [
    { by: "ascending magnitude", sort: "a.properties.mag - b.properties.mag" },
    { by: "descending magnitude", sort: "b.properties.mag - a.properties.mag" },
    { by: "newest", sort: "b.properties.time - a.properties.time" },
    { by: "oldest", sort: "a.properties.time - b.properties.time" },
    {
      by: "closest depth",
      sort: "a.geometry.coordinates[2] - b.geometry.coordinates[2]",
    },
    {
      by: "furthest depth",
      sort: "b.geometry.coordinates[2] - a.geometry.coordinates[2]",
    },
    { by: "closest distance", sort: "a.distance - b.distance" },
    { by: "furthest distance", sort: "b.distance - a.distance" },
  ];

export const firstLoad = (theQuery, fallbackQuery, fallbackFn) => (
  dispatch
) => {
  dispatch({ type: QUAKE_FETCH });
  setTimeout(() => {
    axios
      .get(theQuery)
      .then((res) => {
        dispatch({ type: DISPLAY_QUAKES, quakeData: res.data.features });
      })
      .catch((error) => {
        fallbackFn(fallbackQuery);
      });
  }, 500);
};

export const quakeFetch = (theQuery) => (dispatch) => {
  console.log("in QuakeFetch", theQuery);
  dispatch({ type: QUAKE_FETCH });
  setTimeout(() => {
    axios
      .get(
        `https://earthquake.usgs.gov/fdsnws/event/1/query?format=geojson&limit=100${theQuery}`
      )
      .then((res) => {
        dispatch({ type: DISPLAY_QUAKES, quakeData: res.data.features });
      })
      .catch((err) => {
        console.log("Error Fetching Earthquake Data From USGS: ", err);
        dispatch({ type: QUAKE_FETCH_ERROR });
      });
  }, 500);
};

export const updateSearchParams = (theParams) => (dispatch) => {
  dispatch({ type: UPDATE_SEARCH_PARAMS, params: theParams });
};

export const quakeSort = (sortBy, quakes, state = initialSearchState) => (
  dispatch
) => {
  let myLocation = localStorage.getItem("gps")
    ? JSON.parse(localStorage.getItem("gps"))
    : state;
  // TURF POINT ORDER: longitude, latitude
  var from = turf.point([myLocation.longitude, myLocation.latitude]);

  for (let i = 0; i < sortInfo.length; i++) {
    if (sortBy === sortInfo[i].by) {
      if (
        sortInfo[i].by === "closest distance" ||
        sortInfo[i].by === "furthest distance"
      ) {
        // add distance from quake to user's location into list of quakes
        let quakeDistances = quakes.map((a) => {
          // TURF POINT ORDER: longitude, latitude
          var to = turf.point([
            a.geometry.coordinates[0],
            a.geometry.coordinates[1],
          ]);
          return { ...a, distance: turf.distance(from, to).toFixed(2) };
        });
        sortedQuakes = quakeDistances.sort((a, b) => eval(sortInfo[i].sort));
      } else {
        sortedQuakes = quakes.sort((a, b) => eval(sortInfo[i].sort));
      }
    }
  }
  dispatch({ type: SORT_QUAKES, quakeData: [] });
  dispatch({ type: SORT_QUAKES, quakeData: sortedQuakes });
};

export const setViewport = (viewport) => (dispatch) => {
  let newVP = {
    ...viewport,
    width: "100%",
    height: "100%",
  };

  dispatch({ type: UPDATE_VIEWPORT, viewportInfo: newVP });
};

export const jumpViewport = (long, lat, zoom) => (dispatch) => {
  let newVP = {
    latitude: lat,
    longitude: long,
    width: "100%",
    height: "100%",
    zoom: zoom,
  };

  dispatch({ type: JUMP_VIEWPORT, jumpInfo: newVP });
};

// ***NEWSFEED ACTIONS*** //

export const newsLoad = (newsTopics) => (dispatch) => {
  dispatch({ type: NEWS_FETCH });

  setTimeout(() => {
    let allArticles = [];
    newsTopics.forEach((topic) => {
      axios
        .get(
          `https://api.nytimes.com/svc/search/v2/articlesearch.json?q=${topic}&api-key=${process.env.REACT_APP_NYT_API_TOKEN}`
        )
        .then((res) => {
          let updatedResults = res.data.response.docs.map((article) => {
            return { ...article, topic: topic };
          });
          allArticles = allArticles.concat(updatedResults);

          dispatch({ type: DISPLAY_NEWS, newsData: allArticles });
        })
        .catch((err) => {
          console.log("Error Fetching News Data", err);
          dispatch({ type: NEWS_FETCH_ERROR });
        });
    });
  }, 500);
};
