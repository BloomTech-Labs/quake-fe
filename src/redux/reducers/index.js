// Redux
import { combineReducers } from "redux";

// Reducers
import { quakeReducer } from "./quakeReducer";
import { searchReducer } from "./searchReducer";
import { mapReducer } from "./mapReducer";
import { newsReducer } from "./newsReducer";

// ***** NOTE *****
// When using connect and mapStateToProps, make sure
// you link with this syntax:

// 'prop_name': state.'reducer_name'.'state_value'

// Because we have multiple reducers, you MUST specify
// which state object you are trying to get values from

export const rootReducer = combineReducers({
  quakeReducer,
  searchReducer,
  mapReducer,
  newsReducer,
});
