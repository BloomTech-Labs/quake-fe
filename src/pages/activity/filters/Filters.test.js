import React from "react";

import * as rtl from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom/extend-expect";
import renderer from "react-test-renderer";

import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import { quakeFetch, updateSearchParams } from "../../../redux/actions";

import Filters from "./Filters";
import { getByTestId } from "@testing-library/react";

const mockStore = configureStore([]);

describe("Filters.js Tests", () => {
  let store;
  let filtersComponent;

  beforeEach(() => {
    store = mockStore({
      searchReducer: {
        starttime: "2020-01-01",
        endtime: "2020-01-02",
        minmagnitude: 3,
        maxmagnitude: 10,
        maxradiuskm: 5000,
        latitude: 37.78197,
        longitude: -121.93992,
        placename: "",
      },
    });

    store.dispatch = jest.fn();

    filtersComponent = renderer.create(
      <Provider store={store}>
        <Filters />
      </Provider>
    );
  });

  describe("Filters.js Snapshot", () => {
    it("Filters.js renders with mocked state", () => {
      expect(filtersComponent.toJSON()).toMatchSnapshot();
    });
  });

  describe("Dispatches Actions", () => {
    it("formSubmitCallback() dispatches actions", () => {
      const e = { preventDefault: () => {} };
      jest.spyOn(e, "preventDefault");

      renderer.act(() => {
        filtersComponent.root.findByType("form").props.onSubmit(e);
      });

      expect(e.preventDefault).toHaveBeenCalledTimes(1);
      expect(store.dispatch).toHaveBeenCalledTimes(2); // quakeFetch always dispatches 2 actions
    });
  });
});
