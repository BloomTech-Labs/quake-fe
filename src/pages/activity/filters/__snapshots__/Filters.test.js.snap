// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filters.js Tests Filters.js Snapshot Filters.js renders with mocked state 1`] = `
<dialog
  className="search-menu"
  id="search-menu"
  onClick={[Function]}
>
  <form
    name="filters-form"
    onSubmit={[Function]}
  >
    <fieldset
      className="quick-filters"
      id="quick-filters"
    >
      <legend
        id="quick-filters-label"
      >
        Quick Filters
      </legend>
      <div
        className="quick-filters-flex"
        id="quick-filters-flex"
      >
        <button
          className="quick-filters-button"
          name="https://quakelabs-be-production.herokuapp.com/api/activity/alltime-biggest"
          onClick={[Function]}
        >
          All Time Biggest
        </button>
        <button
          className="quick-filters-button"
          name="https://quakelabs-be-production.herokuapp.com/api/nukes/boom"
          onClick={[Function]}
        >
          Caused by Nukes
        </button>
        <button
          className="quick-filters-button"
          name="https://quakelabs-be-production.herokuapp.com/api/tsunami/splash"
          onClick={[Function]}
        >
          Caused a Tsunami
        </button>
      </div>
    </fieldset>
    <fieldset
      className="advanced-filters"
      id="advanced-filters"
    >
      <legend
        id="advanced-filters-label"
      >
        Advanced Filters
      </legend>
      <fieldset
        className="radius-field"
      >
        <legend>
          Search Radius:
        </legend>
        <label
          htmlFor="kilometers"
        >
          Kilometers
          <input
            data-testid="input-km"
            id="kilometers"
            max="20001.6"
            min="1"
            name="maxradiuskm"
            onChange={[Function]}
            step="0.1"
            type="number"
            value={5000}
          />
        </label>
      </fieldset>
      <fieldset
        className="date-field"
      >
        <legend>
          Date Range:
        </legend>
        <label
          htmlFor="start-date"
        >
          Start
          <input
            id="start-date"
            name="starttime"
            onChange={[Function]}
            type="date"
            value="2020-01-01"
          />
        </label>
        <label
          htmlFor="end-date"
        >
          End
          <input
            id="end-date"
            name="endtime"
            onChange={[Function]}
            type="date"
            value="2020-01-02"
          />
        </label>
      </fieldset>
      <fieldset
        className="mag-field"
      >
        <legend>
          Magnitude Range:
        </legend>
        <label
          htmlFor="start-magnitude"
        >
          Start
          <input
            id="start-magnitude"
            max="10"
            min="0"
            name="minmagnitude"
            onChange={[Function]}
            step="0.1"
            type="number"
            value={3}
          />
        </label>
        <label
          htmlFor="end-magnitude"
        >
          End
          <input
            id="end-magnitude"
            max="10"
            min="0"
            name="maxmagnitude"
            onChange={[Function]}
            step="0.1"
            type="number"
            value={10}
          />
        </label>
      </fieldset>
    </fieldset>
    <button
      className="search-submit-button"
      onClick={[Function]}
      type="submit"
    >
      Search For Activity
    </button>
    <p
      className="close-sort"
      onClick={[Function]}
      role="button"
      type="button"
    >
      Close
    </p>
  </form>
</dialog>
`;
